specification/hugr.md

```diff
@@ -1229,22 +1229,26 @@ Note that in a `NewEdgeSpec` one of `SrcNode` and `TgtNode` is an existing node
 in the hugr and the other is a new node.
 
 The `Replace` method takes as input:
-
-  - a set $S$ of mutually-separated nodes in $\Gamma$;
-  - a partial hugr $G$;
-  - a map $T : \top(G) \to \Gamma \setminus S^*$ whose image consists of container nodes;
-  - a map $B : \bot(G) \to S^\*$ whose image consists of container nodes, such that $B(x)$
-    is separated from $B(y)$ unless $x = y$. Let $X$ be the set of children
-    of nodes in the image of $B$, and $R = S^\* \setminus X^\*$.
+  - the ID of a container node $P$ in $\Gamma$;
+  - a set $S$ of IDs of nodes that are children of $P$
+  - a Hugr $H$ whose root is a node of the same type as $P$. Note this Hugr need not be valid, in that it may be missing:
+      * edges to/from some ports (i.e. it may have unconnected ports)---not just Copyable dataflow outputs, which may occur even in valid Hugrs, but also incoming and/or non-Copyable dataflow ports, and ControlFlow ports,
+      * all children for some container nodes strictly beneath the root (i.e. it may have container nodes with no outgoing hierarchy edges)
+      * some children of the root, for container nodes that require particular children (e.g.
+        $\mathtt{Input}$ and/or $\mathtt{Output}$ if $P$ is a dataflow container, the exit node
+        of a CFG, the required number of children of a conditional)
+  - a map $B$ *from* container nodes in $H$ that have no children *to* container nodes in $S^\*$ none of which is an ancestor of another.
+    Let $X$ be the set of children of nodes in the image of $B$, and $R = S^\* \setminus X^\*$.
   - a list $\mu\_\textrm{inp}$ of `NewEdgeSpec` which all have their `TgtNode`in
-    $G$ and `SrcNode` in $\Gamma \setminus S^*$;
+    $G$ and `SrcNode` in $\Gamma \setminus R$;
   - a list $\mu\_\textrm{out}$ of `NewEdgeSpec` which all have their `SrcNode`in
-    $G$ and `TgtNode` in $\Gamma \setminus S^*$ (and `TgtNode` has an existing
+    $G$ and `TgtNode` in $\Gamma \setminus R$ (and `TgtNode` has an existing
```

Comment:
I'd like to be more precise about the target port/whatever having an existing incoming edge, but that's a little awkward. I'm still wondering about a bigger restructure of edges from being two "lists", but I think the "well-formedness" requirements are so complex anyway that one might as well give up....

One thing I think one might want is to add new edges between nodes in `$\Gamma \setminus R$ i.e. where the source has been removed, the "new" source might also be in the remainder rather than new in the replacement. Forcing all such edges to be "passed through" Input -> Output doesn't work if $P$ isn't a DFG, and doesn't work well for static/order edges even for DFGs.

One possibility might be to allow mu_out to have any SrcNode, although that's tricky because until we've copied $G$ into $\Gamma$ we don't have a consistent notion of ID, hmmm.

A third set of new edges?? There may be call for being able to add arbitrary *new* Order edges, too....

Reply:
Of course one can always use `HugrMut::connect`, I guess we can add this later if we really find a need